src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs

```diff
@@ -6,13 +6,16 @@
 using System.Globalization;
 using System.IO.Pipelines;
 using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
 using Microsoft.AspNetCore.Http.Features;
 using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
 
 internal partial class Http1Connection : HttpProtocol, IRequestProcessor, IHttpOutputAborter
 {
+    internal static ReadOnlySpan<byte> Http2GoAwayProtocolErrorBytes => new byte[17] { 0, 0, 8, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1 };
```

Comment:
GOAWAY has a field for "Additional Debug Data" where we could include an ascii message about the request being sent to an HTTP/1.x only endpoint.

Reply:
Are clients generally good about surfacing "Additional Debug Data"?

I'm more hesitant to take this PR than I am to take #39358 because I think sending a GOAWAY frame makes it less obvious that the client is trying to communicate to an HTTP/1.1 endpoint compared to responding with an HTTP/1.1 400 response which I'd expect most HTTP/2 clients to surface with a protocol error.

If clients aren't great at surfacing the debug data, I'd just keep the clearer log bug continue with the HTTP/1.1 400 response.

Reply:
> Are clients generally good about surfacing "Additional Debug Data"

I don't know, but it would at least show up in network traces and other places visible to the client.

Reply:
My initial testing shows SocketsHttpHandler doesn't use the HTTP/2 GOAWAY at all. It assumes that the reply will be a SETTINGS frame. I'll test more. Might be a bug for the networking team.

Reply:
SocketsHttpHandler treats any frame other than SETTINGS as a protocol error:

https://github.com/dotnet/runtime/blob/dfbae37e91c4744822018dde10cbd414c661c0b8/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L471-L476

This is consistent with the HTTP/2 spec:

https://httpwg.org/specs/rfc7540.html#ConnectionHeader

> The server connection preface consists of a potentially empty SETTINGS frame (Section 6.5) that MUST be the first frame the server sends in the HTTP/2 connection.

Thoughts? Is it always valid for a client to be ready for a GOAWAY?

Reply:
On the error and providing more data, there is the HTTP_1_1_REQUIRED error code. Seems like a better choice than protocol error.

https://httpwg.org/specs/rfc7540.html#rfc.section.7

Reply:
Client issue: https://github.com/dotnet/runtime/issues/63715

Reply:
> On the error and providing more data, there is the HTTP_1_1_REQUIRED error code. Seems like a better choice than protocol error.
> 
> https://httpwg.org/specs/rfc7540.html#rfc.section.7

I don't think it's quite meant for this scenario, but it's close enough.

Reply:
We can send SETTINGS and GOAWAY together, the settings don't have to have anything in them.

Reply:
If the server sends SETTINGS then the client will think a connection was successfully established.

Client handshake sends the preface and server handshake sends SETTINGS.

Reply:
What are you recommending then?

Reply:
Respond to the preface with a GOAWAY. SocketsHttpHandler is going to be updated to read a server GOAWAY at this point. HTTP_1_1_REQUIRED will get displayed in the client exception.

Reply:
I'll test what curl does.

Reply:
curl doesn't care either way (one is old response, other is new HTTP/2 GOAWAY HTTP_1_1_REQUIRED

![image](https://user-images.githubusercontent.com/303201/149411488-897c5115-722e-468a-9583-f28c61c4cd48.png)